src/main/java/hudson/remoting/AsyncFutureImpl.java

```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
         return value;
     }
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        long timeoutMs = unit.toMillis(timeout);
+        // It is not very accurate, but all(proof?) production uses of this class pass seconds.
+        // Test cases pass 1 millisecond and other such fun, but wait(long) is not time-accurate anyway
+        long endWaitTime = System.currentTimeMillis() + timeoutMs;
```

Comment:
:bug: will cause massive issues as the system time changes. You need to use nanotime instead... care for comparison logic though